linkerd/error-metrics/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+#![deny(warnings, rust_2018_idioms)]
+
+use futures::{Future, Poll};
+use indexmap::IndexMap;
+pub use linkerd2_metrics::FmtLabels;
+use linkerd2_metrics::{metrics, Counter, FmtMetrics};
+use std::fmt;
+use std::hash::Hash;
+use std::sync::{Arc, Mutex};
+
+metrics! {
+    request_errors_total: Counter {
+        "The total number of HTTP requests that could not be processed due to a proxy error."
+    }
+}
+
+pub trait LabelError<E> {
+    type Labels: FmtLabels + Hash + Eq;
+
+    fn label_error(&self, error: &E) -> Self::Labels;
+}
+
+#[derive(Debug)]
+pub struct Registry<K: Hash + Eq> {
+    errors: Arc<Mutex<IndexMap<K, Counter>>>,
+}
+
+pub struct RecordErrorLayer<L, K: Hash + Eq> {
+    label: L,
+    errors: Arc<Mutex<IndexMap<K, Counter>>>,
+}
+
+pub struct RecordError<L, K: Hash + Eq, S> {
+    label: L,
+    errors: Arc<Mutex<IndexMap<K, Counter>>>,
+    inner: S,
+}
```

Comment:
What are your thoughts on changing these structs to:

```
pub struct RecordErrorLayer<L, K: Hash + Eq> {
    label: L,
    registry: Registry<K>,
}

pub struct RecordError<L, K: Hash + Eq, S> {
    label: L,
    registry: Registry<K>,
    inner: S,
}
```

When reading through this change, this change is what I expected to see and how I thought about it when understanding the rest of the `impl`s.

Reply:
Alternatively, this could be similar to the `Registry` in #428 where:

```
type Registry<L> = Arc<Mutex<IndexMap<L, Arc<Metrics>>>>;
```

Reply:
Eh, the `Registry` type is really there as a public handle for getting layers and reporting the results... It's a bit confusing for Registry to produce layers but those layers just hold Registries?

This might all be clearer if I split the types up into separate files?

Reply:
Yea it helps, but I think what I'm reacting to is it seems like each `*-metrics` crate has it's own way of implementing what it's concept of a `Registry` is and they are just different enough that I feel like I'm reviewing new impls.

Reviewing this change, the other open #428 metrics change, and the previous `http-metrics` crate change all have a similar purpose but don't really make their metrics recording services the same way.

I'm okay merging this as is, but it may just be something to be aware of that as more metrics are added, see if it makes sense to introduce a reusable "metrics registry".